installers/msi-language/Preset/CustomAction.cs

```diff
@@ -0,0 +1,214 @@
+        {
+            session.Log("un-installing perl file associations");
+            FileAssociation.EnsureAssociationsDeleted(associations());
+
+            session.Log("removing from PATHEXT");
+            var oldPathExt = Environment.GetEnvironmentVariable("PATHEXT", EnvironmentVariableTarget.Machine);
+            var newPathExt = String.Join(";", oldPathExt.Split(';').Where(x => !PathExtensions.Contains(x)).ToArray());
+            if (newPathExt != oldPathExt)
+            {
+                session.Log(string.Format("updating PATHEXT to {0}", newPathExt));
+                Environment.SetEnvironmentVariable("PATHEXT", newPathExt, EnvironmentVariableTarget.Machine);
+            }
+
+            return ActionResult.Success;
+        }
+
+        private FileAssociation[] associations()
+        {
+            return new FileAssociation[] {
+                new FileAssociation {
+                    Extension = ".pl",
+                    ProgId = "Perl.Document",
+                    FileTypeDescription = "Perl Document",
+                    ExecutableFilePath = Path.Combine(this.installPath, "bin", "perl.exe"),
+                },
+                new FileAssociation {
+                    Extension = ".wpl",
+                    ProgId = "WPerl.Document",
+                    FileTypeDescription = "WPerl Document",
+                    ExecutableFilePath = Path.Combine(this.installPath, "bin", "wperl.exe"),
+                }
+            };
+        }
+
+        public ActionResult Install()
+        {
+            session.Log("Install PerlCritic shortcut");
+            var result = PerlCriticShortcut();
+            if (result.Equals(ActionResult.Failure))
+            {
+                session.Log("Could not create Perl Critic shortcut");
+                // Do not fail if we cannot create shortcut
+                return ActionResult.Success;
+            }
+
+            session.Log("install cmd-prompt shortcut");
+            result = CmdPromptShortcut();
+            if (result.Equals(ActionResult.Failure))
+            {
+                session.Log("Could not create Command Prompt shortcut");
+                // Do not fail if we cannot create shortcut
+                return ActionResult.Success;
+            }
+
+            session.Log("installing perl file associations");
+            FileAssociation.EnsureAssociationsSet(associations());
+
+            session.Log("updating PATHEXT");
+            var oldPathExt = Environment.GetEnvironmentVariable("PATHEXT", EnvironmentVariableTarget.Machine);
+            var exts = String.Join(";", oldPathExt.Split(';').Concat(PathExtensions).Distinct().ToArray());
+            if (exts != oldPathExt)
+            {
+                Environment.SetEnvironmentVariable("PATHEXT", exts, EnvironmentVariableTarget.Machine);
+            }
+
+            return ActionResult.Success;
+        }
+
+        private ActionResult PerlCriticShortcut()
```

Comment:
Not related to this line directly however in the `Product.p.wxs` around line 44 we are creating a documentation shortcut that contains a hard coded `Perl` reference. Right now it's just the icon file but we may want to move that shortcut logic into here as well.